inst/include/Rcpp/vector/Matrix.h

```diff
@@ -175,25 +175,25 @@ class Matrix : public Vector<RTYPE, StoragePolicy>, public MatrixBase<RTYPE, tru
 
     template <typename U>
     void fill_diag__dispatch( traits::false_type, const U& u) {
-          Shield<SEXP> elem( converter_type::get( u ) ) ;
-        int n = Matrix::ncol() ;
-        int offset = n +1 ;
-        iterator it( VECTOR::begin()) ;
-        for( int i=0; i<n; i++){
-            *it = ::Rf_duplicate( elem );
-            it += offset;
+        Shield<SEXP> elem( converter_type::get( u ) );
+
+        R_xlen_t bounds = ( Matrix::nrow() < Matrix::ncol() ) ?
+                             Matrix::nrow() : Matrix::ncol();
+        
```

Comment:
I guess one could argue that the ternary operator may be lighter-weight that a function call (which, I bet, ends up inlined).  I use that operator a fair amount for alternates too.   It just seemed more compact...